File: algorithms/2_medium/L/00817_linked_list_components/hashset.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    static int numComponents(ListNode* head, const vector<int>& nums) {
        bool is_prev_valid = false;
        int cnt = 0;
        ListNode* curr = head;
        auto set = unordered_set<int>(nums.cbegin(), nums.cend());

        while (curr != nullptr) {
            const int val = curr->val;
            curr = curr->next;

            if (set.contains(val)) {
                cnt += is_prev_valid ? 0 : 1;
                is_prev_valid = true;
            } else {
                is_prev_valid = false;
            }
        }

        return cnt;
    }
};

File: algorithms/2_medium/L/00817_linked_list_components/hashset.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
use std::collections::HashSet;

impl Solution {
    pub fn num_components(head: Option<Box<ListNode>>, nums: Vec<i32>) -> i32 {
        let mut is_prev_valid = false;
        let mut cnt = 0_i32;
        let mut curr_opt = head;
        let set: HashSet<i32> = nums.into_iter().collect();

        while let Some(mut curr) = curr_opt {
            curr_opt = curr.next.take();
            if set.contains(&curr.val) {
                cnt += if is_prev_valid { 0 } else { 1 };
                is_prev_valid = true;
            } else {
                is_prev_valid = false;
            }
        }

        cnt
    }
}
